gguf-py/gguf/constants.py

```diff
@@ -0,0 +1,469 @@
+from __future__ import annotations
+
+import sys
+from enum import Enum, IntEnum, StrEnum, auto
+from typing import Any, NamedTuple, Type
+
+#
+# constants
+#
+
+GGUF_MAGIC             = 0x46554747 # "GGUF"
+GGUF_VERSION           = 3
+GGUF_DEFAULT_ALIGNMENT = 32
+
+#
+# metadata keys
+#
+
+class GeneralKeys(StrEnum):
+    ARCHITECTURE        : str = "general.architecture"
+    QUANTIZATION_VERSION: str = "general.quantization_version"
+    ALIGNMENT           : str = "general.alignment"
+    NAME                : str = "general.name"
+    AUTHOR              : str = "general.author"
+    URL                 : str = "general.url"
+    DESCRIPTION         : str = "general.description"
+    LICENSE             : str = "general.license"
+    SOURCE_URL          : str = "general.source.url"
+    SOURCE_HF_REPO      : str = "general.source.huggingface.repository"
+    FILE_TYPE           : str = "general.file_type"
+
+class AttentionKeys(StrEnum):
+    HEAD_COUNT       : str = "{arch}.attention.head_count"
+    HEAD_COUNT_KV    : str = "{arch}.attention.head_count_kv"
+    MAX_ALIBI_BIAS   : str = "{arch}.attention.max_alibi_bias"
+    CLAMP_KQV        : str = "{arch}.attention.clamp_kqv"
+    LAYERNORM_EPS    : str = "{arch}.attention.layer_norm_epsilon"
+    LAYERNORM_RMS_EPS: str = "{arch}.attention.layer_norm_rms_epsilon"
+
+class RopeKeys(StrEnum):
+    DIMENSION_COUNT     : str = "{arch}.rope.dimension_count"
+    FREQ_BASE           : str = "{arch}.rope.freq_base"
+    SCALING_TYPE        : str = "{arch}.rope.scaling.type"
+    SCALING_FACTOR      : str = "{arch}.rope.scaling.factor"
+    SCALING_ORIG_CTX_LEN: str = "{arch}.rope.scaling.original_context_length"
+    SCALING_FINETUNED   : str = "{arch}.rope.scaling.finetuned"
+
+class TokenizerKeys(StrEnum):
+    MODEL     : str = "tokenizer.ggml.model"
+    LIST      : str = "tokenizer.ggml.tokens"
+    TOKEN_TYPE: str = "tokenizer.ggml.token_type"
+    SCORES    : str = "tokenizer.ggml.scores"
+    MERGES    : str = "tokenizer.ggml.merges"
+    BOS_ID    : str = "tokenizer.ggml.bos_token_id"
+    EOS_ID    : str = "tokenizer.ggml.eos_token_id"
+    UNK_ID    : str = "tokenizer.ggml.unknown_token_id"
+    SEP_ID    : str = "tokenizer.ggml.seperator_token_id"
+    PAD_ID    : str = "tokenizer.ggml.padding_token_id"
+    HF_JSON   : str = "tokenizer.huggingface.json"
+    RWKV      : str = "tokenizer.rwkv.world"
+
+class LLMKeys(StrEnum):
+    CONTEXT_LENGTH       : str = "{arch}.context_length"
+    EMBEDDING_LENGTH     : str = "{arch}.embedding_length"
+    BLOCK_COUNT          : str = "{arch}.block_count"
+    FEED_FORWARD_LENGTH  : str = "{arch}.feed_forward_length"
+    USE_PARALLEL_RESIDUAL: str = "{arch}.use_parallel_residual"
+    TENSOR_DATA_LAYOUT   : str = "{arch}.tensor_data_layout"
+
+class Keys(NamedTuple):
```

Comment:
This should be Enum, not NamedTuple, and you won't need to specify the types that way. Also, you will then use "Keys.GENERAL", not "KEY.GENERAL".

Reply:
I actually tried and wasn't successful. If you want to change it (or any type annotations, reorganize the examples, etc) you have my blessing to go ahead and just add the commits. 

I'm really going to bed now, I swear.